### PR TITLE
add LEP for feature `handling volume remount by killing the consuming pod`

### PR DESCRIPTION
**Pull request by PhanLe1010**
longhorn/longhorn#1719

### CODE REVIEW

File: enhancements/20201020-support-remount-with-subpath.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+```
+root@worker-1:~# findmnt /dev/longhorn/pvc-1ce69c7e-90ce-41ce-a88e-8b968d9a8ff9 
+TARGET                                                                                                                                  SOURCE                                                        FSTYPE OPTIONS
+/var/lib/kubelet/pods/0429305b-faf1-4668-aaed-139a2cf4989c/volumes/kubernetes.io~csi/pvc-1ce69c7e-90ce-41ce-a88e-8b968d9a8ff9/mount     /dev/longhorn/pvc-1ce69c7e-90ce-41ce-a88e-8b968d9a8ff9        ext4   rw,rela
+/var/lib/kubelet/pods/0429305b-faf1-4668-aaed-139a2cf4989c/volume-subpaths/pvc-1ce69c7e-90ce-41ce-a88e-8b968d9a8ff9/nginx/0             /dev/longhorn/pvc-1ce69c7e-90ce-41ce-a88e-8b968d9a8ff9[/html] ext4   rw,rela
+```
+We can identify which mount point is subpath and which mount point mounts to the root of the volume's filesystem.
+Then, Longhorn can remount the mount point which mounts to the root of the volume's filesystem.
+
+### #2 Kill the consuming pod
```
Comment (yasker) [inline enhancements/20201020-support-remount-with-subpath.md:36]: `Delete the workload pod`.

We don't call it `consuming` pod before. Stick with our language.

File: enhancements/20201020-support-remount-with-subpath.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+Instead of manually finding and remounting all mount points of the volume, we kill the pod so that Kubernetes handles those works.
```
Comment (yasker) [inline enhancements/20201020-support-remount-with-subpath.md:38]: Need to clarify that what do we expect Kubernetes to do.

File: enhancements/20201020-support-remount-with-subpath.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+I would like to choose approach #2, `kill the consuming pod` because it is cleaner.
+Manually doing remount is cumbersome and leaves duplicated mount points on the host.
+
+### User Stories
+
+#### Story 1
+
+Users use subpath in PVC that is bounded to Longhorn volume.
+When the network goes bad, the volume becomes faulty (if there is no local replica).
+Longhorn auto salvages the volume when the network is good.
```
Comment (yasker) [inline enhancements/20201020-support-remount-with-subpath.md:52]: `when the network comes back`.

File: enhancements/20201020-support-remount-with-subpath.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+When the network goes bad, the volume becomes faulty (if there is no local replica).
+Longhorn auto salvages the volume when the network is good. 
+Then Longhorn auto remount, the current remount logic doesn't support subpath.
+
+### User Experience In Detail
+
+When users deploy workload using controller such as deployment, statefulset, or daemonset, 
+They are assured that volume gets reattached and remounted in case it is detached unexpectedly. 
+
+What about the orphan pods (with no controller)? Users have to manually kill and recreate orphan pods.
```
Comment (yasker) [inline enhancements/20201020-support-remount-with-subpath.md:60]: I am not sure those are called `orphan pods`. You can say `a pod without a controller`.

**Review by yasker — CHANGES_REQUESTED** [review]

File: enhancements/20201020-support-remount-with-subpath.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
+# Handling Volume Remount by Killing the Consuming Pod
```
[the remark this answers is not in the document]
Comment (PhanLe1010) [reply enhancements/20201020-support-remount-with-subpath.md]: Thank you. Fixed

File: enhancements/20201020-support-remount-with-subpath.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+## Summary
+
+The current implementation of the remount feature doesn't work when the consuming pod uses subpath.
```
[the remark this answers is not in the document]
Comment (PhanLe1010) [reply enhancements/20201020-support-remount-with-subpath.md]: Thank you. Fixed

File: enhancements/20201020-support-remount-with-subpath.md (hunk 1 above)
Comment (PhanLe1010) [reply enhancements/20201020-support-remount-with-subpath.md]: Thank you. Fixed

File: enhancements/20201020-support-remount-with-subpath.md (hunk 2 above)
Comment (PhanLe1010) [reply enhancements/20201020-support-remount-with-subpath.md]: Thank you. Fixed

File: enhancements/20201020-support-remount-with-subpath.md (hunk 3 above)
Comment (PhanLe1010) [reply enhancements/20201020-support-remount-with-subpath.md]: Thank you. Fixed

File: enhancements/20201020-support-remount-with-subpath.md (hunk 4 above)
Comment (PhanLe1010) [reply enhancements/20201020-support-remount-with-subpath.md]: Thank you. Fixed

File: enhancements/20201020-support-remount-with-subpath.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+1. Kill the replica instance manager pod on `node-1` one more time. 
+1. Wait for volume to become healthy, kill the replica instance manager pod on `node-1` one more time. 
+1. in a 2 minutes retry loop: 
+   Exec into the `pod-1`, run `ls /mnt/html`. 
+   Verify the file `test_data.txt` exists.
+1. Delete `pod-1`.
+1. in a 2 minutes retry loop: 
+   Exec into the `pod-1`, run `ls /mnt/html`. 
+   Verify the file `test_data.txt` exists.
+
```
Comment (yasker) [inline enhancements/20201020-support-remount-with-subpath.md:108]: Add the engine instance manager part.

**Review by yasker — CHANGES_REQUESTED** [review]
Need to change the filename too.

File: enhancements/20201020-support-remount-with-subpath.md (hunk 7 above)
Comment (PhanLe1010) [reply enhancements/20201020-support-remount-with-subpath.md]: Thank you. Fixed

**Review by yasker — APPROVED** [review]

**Pull request merged**